File: dotnet/Nitro/Geometry/DoubleVector2.cs
namespace Nitro.Geometry
{
    public class DoubleVector2
    {
        public double X;

        public double Y;

        public DoubleVector2()
        {
            X = Y = 0;
        }

        public DoubleVector2(double x, double y)
        {
            X = x;
            Y = y;
        }

        public DoubleVector2(DoubleVector2 vector)
        {
            X = vector.X;
            Y = vector.Y;
        }

        /// <summary>
        /// Returns the length of this vector
        /// </summary>
        public double Length()
        {
            return Math.Sqrt(X * X + Y * Y);
        }

        /// <summary>
        /// Returns |this| * |this|. Performs better than Length() and is sufficient
        /// for comparisions when e.g. sorting vectors by length.
        /// </summary>
        public double LengthSq()
        {
            return X * X + Y * Y;
        }

        /// <summary>
        /// Performs the dot product
        /// </summary>
        public double Dot(DoubleVector2 a)
        {
            return this.X * a.X + this.Y * a.Y;
        }

        /// <summary>
        /// Orthographic Vector
        /// </summary>
        public DoubleVector2 Orthographic
        {
            get
            {
                return new DoubleVector2(-this.Y, this.X);
            }
        }

        /// <summary>
        /// Checks if two vectors are colinear ("parallel")
        /// </summary>
        /// <returns>true, if this and vector a are colinear.</returns>
        public bool IsColinear(DoubleVector2 a)
        {
            DoubleVector2 v1 = new DoubleVector2(this);
            DoubleVector2 v2 = new DoubleVector2(a);

            v1.Normalize();
            v2.Normalize();

            return (v1 - v2).LengthSq() < 0.001;
        }

        /// <summary>
        /// Calculates the Angle included between this and the given vector
        /// </summary>
        /// <returns>Angle in radians</returns>
        public double Angle(DoubleVector2 vector)
        {
            double numerator = this.Dot(vector);
            double denominator = this.Length() * vector.Length();

            return Math.Acos(numerator / denominator);
        }

        /// <summary>
        /// Calculates distance from this point to the given one
        /// </summary>
        /// <returns>Distance</returns>
        public double DistanceTo(DoubleVector3 vector)
        {
            return Math.Sqrt((this.X - vector.X) * (this.X - vector.X) +
                (this.Y - vector.Y) * (this.Y - vector.Y));
        }

        public override string ToString()
        {
            return String.Format("{0:0.00000}, {1:0.00000}", X, Y);
        }

        public override bool Equals(object? obj)
        {
            if (obj is not DoubleVector2)
                return false;

            DoubleVector2 vector = (DoubleVector2)obj;
            return vector.X == X && vector.Y == Y;
        }

        public override int GetHashCode()
        {
            return X.GetHashCode() + Y.GetHashCode();
        }

        public static DoubleVector2 operator +(DoubleVector2 a, DoubleVector2 b)
        {
            return new DoubleVector2(a.X + b.X, a.Y + b.Y);
        }

        public static DoubleVector2 operator -(DoubleVector2 a)
        {
            return new DoubleVector2(-a.X, -a.Y);
        }

        public static DoubleVector2 operator -(DoubleVector2 a, DoubleVector2 b)
        {
            return new DoubleVector2(a.X - b.X, a.Y - b.Y);
        }

        public static DoubleVector2 operator *(DoubleVector2 a, double factor)
        {
            return new DoubleVector2(a.X * factor, a.Y * factor);
        }

        public static DoubleVector2 operator *(double factor, DoubleVector2 a)
        {
            return new DoubleVector2(a.X * factor, a.Y * factor);
        }

        public static double operator *(DoubleVector2 a, DoubleVector2 b)
        {
            return a.Dot(b);
        }

        public static DoubleVector2 operator /(DoubleVector2 a, double factor)
        {
            return new DoubleVector2(a.X / factor, a.Y / factor);
        }

        public static bool operator ==(DoubleVector2? a, DoubleVector2? b)
        {
            return a?.X == b?.X && a?.Y == b?.Y;
        }

        public static bool operator !=(DoubleVector2? a, DoubleVector2? b)
        {
            return !(a == b);
        }

        /// <summary>
        /// Performs in-place addition
        /// </summary>
        public DoubleVector2 Add(DoubleVector2 a)
        {
            this.X += a.X;
            this.Y += a.Y;

            return this;
        }

        /// <summary>
        /// Performs in-place scalar multiplication
        /// </summary>
        public DoubleVector2 Multiply(double scalar)
        {
            this.X *= scalar;
            this.Y *= scalar;

            return this;
        }

        /// <summary>
        /// Normalizes this vector. Vector remains unchanged in case length is 0.
        /// </summary>
        public DoubleVector2 Normalize()
        {
            double length = Length();

            if (length == 0)
                return this;

            X /= length;
            Y /= length;

            return this;
        }
    }
}

File: web/webmap/src/geometry/DoubleVector2.ts
import { DoubleVector3 } from "./DoubleVector3";

export class DoubleVector2 {
    constructor(public x: number = 0, public y: number = 0) {
    }

    public length(): number {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }

    public lengthSq(): number {
        return this.x * this.x + this.y * this.y;
    }

    public dot(vec: DoubleVector2): number {
        return this.x * vec.x + this.y * vec.y;
    }

    public getOrthographic(): DoubleVector2 {
        return new DoubleVector2(-this.y, this.x);
    }

    public isColinear(a: DoubleVector2): boolean {
        const v1 = new DoubleVector2(this.x, this.y);
        const v2 = new DoubleVector2(a.x, a.y);

        v1.normalize();
        v2.normalize();

        return v1.subtract(v2).lengthSq() < 0.001;
    }

    public angleBetween(vector: DoubleVector2): number {
        const numerator = this.dot(vector);
        const denominator = this.length() * vector.length();

        return Math.acos(numerator / denominator);
    }

    public distanceTo(vector: DoubleVector2): number {
        return Math.sqrt(
            (this.x - vector.x) * (this.x - vector.x) +
                (this.y - vector.y) * (this.y - vector.y)
        );
    }

    public toString(): string {
        return `${this.x.toFixed(5)}, ${this.y.toFixed(5)}`;
    }

    public equals(obj?: DoubleVector2): boolean {
        if (!obj) 
            return false;

        const vector = obj as DoubleVector2;
        return vector.x === this.x && vector.y === this.y;
    }

    public static add(a: DoubleVector2, b: DoubleVector2): DoubleVector2 {
        return new DoubleVector2(a.x + b.x, a.y + b.y);
    }

    public static subtract(a: DoubleVector2, b: DoubleVector2): DoubleVector2 {
        return new DoubleVector2(a.x - b.x, a.y - b.y);
    }

    public static multiply(a: DoubleVector2, factor: number): DoubleVector2 {
        return new DoubleVector2(a.x * factor, a.y * factor);
    }

    public static divide(a: DoubleVector2, factor: number): DoubleVector2 {
        return new DoubleVector2(a.x / factor, a.y / factor);
    }

    public static equals(a?: DoubleVector2, b?: DoubleVector2): boolean {
        return a?.x === b?.x && a?.y === b?.y;
    }

    public add(a: DoubleVector2): DoubleVector2 {
        this.x += a.x;
        this.y += a.y;

        return this;
    }

    public multiply(scalar: number): DoubleVector2 {
        this.x *= scalar;
        this.y *= scalar;

        return this;
    }

    public normalize(): DoubleVector2 {
        const length = this.length();

        if (length === 0) return this;

        this.x /= length;
        this.y /= length;

        return this;
    }

    public subtract(a: DoubleVector2): DoubleVector2 {
        return new DoubleVector2(this.x - a.x, this.y - a.y);
    }

    public toDoubleVector3(): DoubleVector3 {
        return new DoubleVector3(this.x, this.y, 0);
    }
}
